tests/test_charms.py

```diff
@@ -0,0 +1,251 @@
+from pathlib import Path
+from random import choices
+from string import ascii_lowercase
+from subprocess import check_output, run
+from time import sleep
+from urllib.parse import urlparse
+
+import pytest
+import yaml
+from selenium.common.exceptions import JavascriptException, WebDriverException
+from selenium.webdriver.common.by import By
+from selenium.webdriver.firefox.options import Options
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.ui import WebDriverWait
+from seleniumwire import webdriver
+
+CONTROLLER_METADATA = yaml.safe_load(Path("charms/jupyter-controller/metadata.yaml").read_text())
+UI_METADATA = yaml.safe_load(Path("charms/jupyter-ui/metadata.yaml").read_text())
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test):
+    run(["juju", "switch", ops_test.model_full_name])
```

Comment:
This and the other calls to `run` in this file should probably switch to `await ops_test.run(...)` instead of the one provided by `subprocess`, as that's blocking and `ops_test.run` isn't.

Note that instead of a list of args, `ops_test.run(...)` expects `*args`. I.e. this line would be `await ops_test.run('juju', 'switch', ops_test.model_full_name)`